### ISSUE TITLE
NamedArguments: Ignore when argument values are the same as the parameter names

### ISSUE THREAD

**Issue by henrikbarium** [detekt/detekt#4591]
## Expected Behavior of the rule
The idea is to introduce a new setting for `NamedArguments` called something like `ignoreSimpleArgumentValues` (preferably a better name :-) ). When enabled, the rule would not require named arguments in cases where the argument values are variables whose name are the same as the corresponding parameters.

Example:
```kotlin
import java.time.LocalDate

// Obviously made-up function which has parameter names which are roughly as long as the ones in our codebase
fun processCandidateForMission(
    birthName: String,
    dateOfBirth: LocalDate,
    fitessForPurpose: Float,
    willingToParticipate: Boolean,
    licensedToKill: Boolean
): Unit = TODO()
fun calculateFitness(): Float = TODO()

// 1) Should be reported - not obvious what the values are
processCandidateForMission("John Doe", LocalDate.of(2000, 1, 1), calculateFitness(), true, false)

// 2) Should not be reported - we have added the required parameter names which helps a lot in readability
processCandidateForMission(
    birthName = "John Doe",
    dateOfBirth = LocalDate.of(2000, 1, 1),
    fitessForPurpose = calculateFitness(),
    willingToParticipate = true,
    licensedToKill = false
)

// Let's define variables matching the parameters.
// We often do this in our codebase, though the expressions are often more complex.
val birthName = "John Doe"
val dateOfBirth = LocalDate.of(2000, 1, 1)
val fitessForPurpose = calculateFitness()
val willingToParticipate = true
val licensedToKill = false

// 3) Is currently reported but should not IMO - adding parameter names will not add anything
processCandidateForMission(birthName, dateOfBirth, fitessForPurpose, willingToParticipate, licensedToKill)

// 4) Parameter names are just verbose and not really helpful - this duplication is what I want to avoid
//    It turns a simple one-line function call without any complexity into an ugly seven-line monster of 
//    verbosity and duplication. ;-) 
processCandidateForMission(
    birthName = birthName,
    dateOfBirth = dateOfBirth,
    fitessForPurpose = fitessForPurpose,
    willingToParticipate = willingToParticipate,
    licensedToKill = licensedToKill
)
```

## Context
I want to enable the `NamedArguments` rule in our codebase. In many cases it helps making code more readable. However, we have many cases where we define local variables with the same names as the parameters, and then use these variables as parameter values. In those cases adding named parameters is just redundant and makes the code _less_ readable.

The idea is similar to where IntelliJ automatically annotates the code with "virtual" parameter names and/or type declarations when it considers it helpful. With the example above, IntelliJ annotates call (1) with parameter names but it does not annotate (3), presumably because it would not add any information.

I do appreciate that this may be a special case, but at least in our codebase it is common. :-)

**Comment by cortinico** [comment]
> Parameter names are just verbose and not really helpful

Named arguments actually protect you against parameter re-ordering so they're definitely helpful

**Comment by henrikbarium** [comment]
To be clear, when I said that parameter names are "just verbose and not really helpful" I ment only in that particular case when the parameter names and variable names were exactly the same. In general, parameter names can of course be useful.

With the rule behavior that I suggest the parameter re-ordering protection would still be in place. Since the rule would only allow named parameters to be skipped if the argument variable name matches exactly, the rule would fail the code if a parameter changed place.

**Comment by cortinico** [comment]
> With the rule behavior that I suggest the parameter re-ordering protection would still be in place. Since the rule would only allow named parameters to be skipped if the argument variable name matches exactly, the rule would fail the code if a parameter changed place.

I'm not sure I got your point. Could you make some code examples?

Specifically in a scenario like:

```kotlin
fun login(password: String, username: String) { /* ... */ }

login(username = username, password = password)
```

Are you suggesting that the rule should flag those two named parameters because the name matches exactly?

**Comment by henrikbarium** [comment]
Given your `login` function, the idea is that the rule would flag:

```kotlin
login("myusername", "mypassword") // No named parameters
login(someFunction(), some + expression) // No named parameters
login(somevariable, someothervariable) // No named parameters and variable names does not match parameter names
```

But it would not flag:

```kotlin
login(username, password) // Ok without named parameters because variable names matches parameter names
login(username = "myusername", password = "mypassword") // Ok because named parameters are used
login(username = username, password = password) // Ok because named parameters are used (though a bit redundant)
```

**Comment by cortinico** [comment]
I see, thanks for giving some examples.
Yes we could add a parameter to add this extra behavior.

> login(password, username) // Ok without named parameters because variable names matches parameter names

I just wanted to point out that not using named parameters here, is a bad idea IMHO. You could argue that detekt will catch if parameter names changed so a parameter reordering won't affect you. On the other hand, we should strive to flag code that is unsafe and using named parameters here (i.e. `login(username = username, password = password)` should be preferred).

Therefore if we add this config (say `ignoreArgumentsMatchingNames`), it should be disabled by default.

**Comment by henrikbarium** [comment]
I see your point. For my part, we're using multiple plugins and tools to verify that our code base is correct (detekt, ktlint, archunit, various dependency checkers, etc), so I guess I've gotten quite used to rely on these tools.

If such an option was added, I would definitely use it. :-) If you think adding it is a step in the wrong direction, I'll accept that too.

**Comment by BraisGabin** [comment]
I think that this config have sense. I agree that if the names are exactly the same there is no need to add them.

But I would talk first ahout some edge cases. Given this function:

```kotlin
fun processCandidateForMission(
    birthName: String,
    dateOfBirth: LocalDate,
    fitessForPurpose: Float,
    willingToParticipate: Boolean,
    licensedToKill: Boolean
): Unit = TODO()
```

Should this be flagged?

```kotlin
processCandidateForMission(
    "Brais", // the first one is a value
    dateOfBirth,
    fitessForPurpose,
    willingToParticipate,
    licensedToKill
)
```

```kotlin
processCandidateForMission(
    birthName,
    dateOfBirth,
    fitessForPurpose2, // one parameter is slightly different
    willingToParticipate,
    licensedToKill
)
```


```kotlin
something.let {
  processCandidateForMission(
      birthName,
      it.dateOfBirth, // use it.
      fitessForPurpose,
      willingToParticipate,
      licensedToKill
  )
}
```

```kotlin
something.apply {
  processCandidateForMission(
      birthName,
      this.dateOfBirth, // use this.
      fitessForPurpose,
      willingToParticipate,
      licensedToKill
  )
}
```

To me both cases should be flagged. and only allow when ALL the parameters are EXACTLY the same. But, to me this should be fine:

```kotlin
something.apply {
  processCandidateForMission(
      birthName,
      dateOfBirth, // even if we are accesing a property from `something` instead of a local variable
      fitessForPurpose,
      willingToParticipate,
      licensedToKill
  )
}
```

**Comment by henrikbarium** [comment]
I agree with you. To make the rule consistent and easy to understand I think all parameters should have to be exactly the same. And I also think the last example would be ok.

**Comment by johngray1965** [comment]
> To be clear, when I said that parameter names are "just verbose and not really helpful" I ment only in that particular case when the parameter names and variable names were exactly the same.

I had function with arguments for pageWidth, pageHeight, renderWidth, renderHeight and a few others.   We consistently passed the swapped page* and render*.  The parameter names were named same.   But that didn't really prevent us from getting it wrong.

**Comment by BraisGabin** [comment]
I don't fully get your point. But if you want any change on detekt you should create a new issue explaining your use case. You can link this issue if you think it is relevant.
